lang/cs/Org.Apache.REEF.Common/metrics/MetricsSystem/MetricsSystem.cs

```diff
@@ -0,0 +1,307 @@
+    /// whole flow of metrics from sources to consumers/sinks. Snapshot is taken either at 
+    /// fixed configurable intervals or by an immediate request from the user via 
+    /// <see cref="PublishMetricsNow"/>. This is a thread-safe class.
+    /// </summary>
+    internal sealed class MetricsSystem : IMetricsSystem
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(MetricsSystem));
+
+        private readonly Dictionary<string, MetricsSourceHandler> _sources =
+            new Dictionary<string, MetricsSourceHandler>();
+
+        private readonly Dictionary<int, SinkHandler> _sinks = new Dictionary<int, SinkHandler>();
+        private readonly SinkHandlerParameters _sinkParameters;
+        private int _currentSinkCount = 0;
+        private readonly Timer _timer;
+        private readonly IMetricsCollectorMutable _metricsCollector;
+        private readonly IList<SourceSnapshot> _sourceSnapshots = new List<SourceSnapshot>();
+        private bool _shutDown = false;
+        private readonly object _lock = new object();
+        private readonly IMetricsFilter _sourceFilter;
+
+        /// <summary>
+        /// Injectable constructor
+        /// </summary>
+        /// <param name="sinkParameters">Parameters for creating sink handler.</param>
+        /// <param name="periodicTimerInMs">Periodic time after which snapshot will be taken.</param>
+        /// <param name="getUnchangedMetrics">Whether to get even unchanged metrics from the sources.</param>
+        /// <param name="metricsCollector">The metrics collector used to take snapshot.</param>
+        /// <param name="sourceFilter">Filter that allows metrics system to skip some particular sources.</param>
+        [Inject]
+        private MetricsSystem(SinkHandlerParameters sinkParameters,
+            [Parameter(typeof(MetricsSystemPeriodicTimer))] int periodicTimerInMs,
+            [Parameter(typeof(MetricsSystemGetUnchangedMetrics))] bool getUnchangedMetrics,
+            IMetricsCollectorMutable metricsCollector,
+            IMetricsFilter sourceFilter)
+        {
+            _sinkParameters = sinkParameters;
+            _timer = new Timer(periodicTimerInMs);
+            _timer.Elapsed += (sender, e) => PublishMetrics(getUnchangedMetrics);
+            _metricsCollector = metricsCollector;
+            _sourceFilter = sourceFilter;
+        }
+
+        /// <summary>
+        /// Subscribes a new consumer/sink with the system. Throws exception if sink 
+        /// is already registered or system is in shutdown mode. 
+        /// </summary>
+        /// <param name="observer">The consumer/sink.</param>
+        /// <returns>A reference to interface that allows sink to stop receiving metrics.</returns>
+        public IDisposable Subscribe(IObserver<IMetricsRecord> observer)
+        {
+            lock (_lock)
+            {
+                if (!_shutDown)
```

Comment:
Please invert this `if` and the next one, so that it's clear that exceptions are thrown in exceptional scenarios and are not the default behavior